pkg/reconciler/rolloutorchestrator/rolloutorchestrator.go

```diff
@@ -0,0 +1,395 @@
+}
+
+// IsStageScaleUpReady decides whether the scaling up has completed or on the way for the current stage, based
+// on the revision and the spa(StagePodAutoscaler).
+func IsStageScaleUpReady(spa *v1.StagePodAutoscaler, revision *v1.TargetRevision) bool {
+	if spa.Status.DesiredScale == nil || spa.Status.ActualScale == nil {
+		return false
+	}
+	min := getMinScale(revision)
+	max := getMaxScale(revision)
+	if revision.TargetReplicas == nil {
+		// For revision scaling up without TargetReplicas, it means this revision will be assigned 100% of the traffic.
+		return actualScaleBetweenMinMax(spa, min, max)
+	}
+
+	// There are two modes to scale up and down the replicas of the revisions:
+	// 1. No traffic. Knative Service specifies both min and max scales for the revision. We need to control the
+	// StageMinScale and StageMaxScale to make sure the replicas increase or decrease. In this case, we need to
+	// precisely make sure both DesiredScale and ActualScale are equal to or greater than TargetReplicas to determine
+	// scaling up phase is over. TargetReplicas is no larger than minScale, because revision runs at the number of
+	// minScale. We need to first scale up the new revision, make sure it run at the correct number, and scale down the
+	// old revision. Shifting traffic does not change anything in terms of the number of replicas.
+	// 2. Traffic driven. In this case, TargetReplicas is larger than minScale and lower than or equal to maxScale
+	// for the revision. We need to change the number of the replicas by shifting the traffic. As long as we know the
+	// new revision is on the way of scaling up, we are able to start the scaling down phase as well.
+	if *spa.Status.DesiredScale >= *revision.TargetReplicas && *spa.Status.ActualScale >= *revision.TargetReplicas {
+		// This is for the first mode.
+		return true
+	}
+	if *spa.Status.DesiredScale >= min && *spa.Status.DesiredScale == *spa.Status.ActualScale {
+		// This is for the second mode.
+		return true
+	}
+
+	return false
+}
+
+// IsStageScaleDownReady decides whether the scaling down has completed for the current stage, based
+// on the revision and the spa(StagePodAutoscaler).
+func IsStageScaleDownReady(spa *v1.StagePodAutoscaler, revision *v1.TargetRevision) bool {
+	if spa.Status.DesiredScale == nil || spa.Status.ActualScale == nil {
+		return false
+	}
+	if revision.TargetReplicas == nil {
+		// For revision scaling up without TargetReplicas, it means this revision will be assigned 0% of the traffic.
+		max := getMaxScale(revision)
+		return bothValuesUnderTargetValue(*spa.Status.DesiredScale, *spa.Status.ActualScale, max)
+	}
+	return bothValuesUnderTargetValue(*spa.Status.DesiredScale, *spa.Status.ActualScale, *revision.TargetReplicas)
+}
+
+func bothValuesUnderTargetValue(desire, actual, target int32) bool {
+	return desire <= target && actual <= target
```

Comment:
Do we need to check the desiredScale as it could be traffic driven and possibly be much larger than actualScale

Reply:
Even if it is driven by the traffic, the actualScale will reach the value of desiredScale ultimately. The desiredScale will become stabilized, when the incoming traffic remains at a stable level.

The actualScale is always smaller than or equal to desiredScale, but it will always try to become equal to desiredScale.

Reply:
Not necessarily as we have resource quota for the upper limit. Say the quota is 10 cpu and autoscaler is asking for 20 replicas with 20 cpu, so it will never reach to the desired scale.